components/popover/popover-mixin.js

```diff
@@ -794,33 +831,55 @@ export const PopoverMixin = superclass => class extends superclass {
 
 		const pointerRect = pointer.getBoundingClientRect();
 
-		if (this._preferredPosition.span !== 'all') {
-			const xAdjustment = Math.min(20 + ((pointerRotatedLength - pointerLength) / 2), (openerRect.width - pointerLength) / 2);
-			if (!this._rtl) {
-				if (this._preferredPosition.span === 'end') {
-					position.left = openerRect.left + xAdjustment;
+		if (this._location === 'block-end' || this._location === 'block-start') {
+
+			if (this._preferredPosition.span !== 'all') {
+				const xAdjustment = Math.min(20 + ((pointerRotatedLength - pointerLength) / 2), (openerRect.width - pointerLength) / 2);
+				if (!this._rtl) {
+					if (this._preferredPosition.span === 'end') {
+						position.left = openerRect.left + xAdjustment;
+					} else {
+						position.right = (openerRect.right * -1) + xAdjustment;
+					}
 				} else {
-					position.right = (openerRect.right * -1) + xAdjustment;
+					if (this._preferredPosition.span === 'end') {
+						position.right = window.innerWidth - openerRect.right + xAdjustment - getScrollbarWidth();
+					} else {
+						position.left = (window.innerWidth - openerRect.left - xAdjustment - getScrollbarWidth()) * -1;
+					}
 				}
 			} else {
-				if (this._preferredPosition.span === 'end') {
-					position.right = window.innerWidth - openerRect.right + xAdjustment;
+				if (!this._rtl) {
+					position.left = openerRect.left + ((openerRect.width - pointerRect.width) / 2);
 				} else {
-					position.left = (window.innerWidth - openerRect.left - xAdjustment) * -1;
+					position.right = window.innerWidth - openerRect.left - ((openerRect.width + pointerRect.width) / 2) - getScrollbarWidth();
 				}
 			}
-		} else {
-			if (!this._rtl) {
-				position.left = openerRect.left + ((openerRect.width - pointerRect.width) / 2);
+
+			if (this._location === 'block-start') {
+				position.bottom = window.innerHeight - openerRect.top + this._offset - 8;
 			} else {
-				position.right = window.innerWidth - openerRect.left - ((openerRect.width + pointerRect.width) / 2);
+				position.top = openerRect.top + openerRect.height + this._offset - 7;
+			}
+
+		} else if (this._location === 'inline-end' || this._location === 'inline-start') {
+
+			position.top = openerRect.top + (openerRect.height / 2) - 9;
```

Comment:
nit: a comment on what the numbers like 9 or 7 are used for could be useful

Reply:
I managed to get rid of the arbitrary 9, but adding comments for the others.